### DISCUSSION TITLE
Is it possible to subclass DTOs?

### DISCUSSION THREAD

**Discussion by simon-lund** [litestar-org/litestar#2719] [Q&A] [answered]
I would like to define subclassed DTOs mapping to my internal data structure schema. There I use a BaseRecord, which is subclassed by various records:

```python
class BaseRecord(BaseModel):
    id: str
    updated: str
    created: str

class UserRecord(BaseRecord):
    name: str
    email: str
    password: str

...
```

Now I want to create a BaseDTO that includes a configuration for the BaseRecord. 
Here I'd like to rename 'update' and 'created' fields to 'updatedAt' and 'createdAt' respectively, instead of configuring this renaming for each individual DTO. 

In addition, I would like to be able to update this configuration in subclasses, for example, to exclude fields.

**Comment by peterschutt** [comment]
I assume that `UserRecord` is meant to inherit from `BaseRecord` in your example?

Is there a specific error you are encountering?

I believe something like this should work, but its just off the top of my head so might be errors in it:

```py
ModelT = TypeVar("ModelT", bound=BaseRecord)

class BaseRecordDTO(PydanticDTO[ModelT]):
   config = DTOConfig(rename={"updated": "updatedAt", "created": "createdAt"})

UserRecordDTO = BaseRecordDTO[UserRecord]
```

**Reply by simon-lund** [reply]
Thank you. Yes, wrong subclass ^^ sorry!

Does this still work if one needs to update the config, for example, exclude a field in UserRecordDTO?

```python
class UserRecordDTO(BaseRecordDTO[UserRecord]):
    config = DTOConfig(exclude="internal_xy"))
```

**Reply by peterschutt** [reply]
No - it is not obvious to me how all of the configuration parameters should be merged. Open to suggestions.

One thought off the top of my head is we have a callback method on AbstractDTO that handles the case where a 2nd config object is found in the mro of the DTO type.. something like:

```py
class AbstractDTO:
    ...

    @staticmethod
    def merge_configs(old: DTOConfig, new: DTOConfig) -> DTOConfig:
        return new
```

Default would just return the latest one, which would preserve current overwrite behavior, but then you could do something like:

```py
class MyDTO(PydanticDTO[...]):
    ...

    @staticmethod
    def merge_configs(old: DTOConfig, new: DTOConfig) -> DTOConfig:
        new.rename.update(old.rename)
        return new
```

**Reply by simon-lund** [reply]
I like this idea very much :+1:

**Reply by peterschutt** [reply]
@litestar-org/members should we do something like this?

**Reply by peterschutt** [reply, marked as answer]
https://github.com/litestar-org/litestar/issues/2760

**Comment by vkcku** [comment]
I'm not against this but I still have some doubts regarding the merging of the config. @peterschutt, the current method you're proposed works if the user subclasses from only one parent DTO. What if they subclass from multiple DTOs? I can see a scenario where users would create multiple "base" DTOs with various configurations to then be used as mixins to form the final DTO they want.

**Reply by peterschutt** [reply]
Can you give me an example of what subclassing from multiple dtos looks like, and some idea of the problem that solves?
